### PR TITLE
hydra: first version

### PR DESCRIPTION
**Pull request by PascalNehlsen**
- completed code 
- added hydra readme 
- added docstrings 
- added type hints

### CODE REVIEW

File: .gitignore
Comment (spmse) [inline .gitignore:1]: why did you delete this file? Should be kept here, as it is already fine

File: README.md
Comment (spmse) [inline README.md:1]: Why did you delete the Top-Level README? Should be kept to provide an overview about the repository, its contents, and thus your work!

You can (and should) add more READMEs for each of the subfolders where own tasks live in --> each specific README describes the folders content and serves as an entry point for a repository visitor, so should the root README do, too. Please re-add the file (with some slight adjustments)

File: module-5/hydra/README.md [hunk 1]
```diff
@@ -1,18 +1,130 @@
-# My `hydra` tool
+# My `hydra` Tool
```
Comment (spmse) [inline module-5/hydra/README.md:1]: 😅

File: module-5/hydra/README.md [hunk 2]
```diff
@@ -1,18 +1,130 @@
...
+## Features
 
-## Usage examples
+This implementation covers the following features/options:
 
-TBD
\ No newline at end of file
+- **SSH Brute-Force Attack**: Attempts to crack SSH passwords by trying all combinations within a given length and character set.
+- **Dictionary Attack**: Uses a provided wordlist to attempt to find the correct password.
+- **Configurable Parameters**: Allows customization of minimum and maximum password lengths, as well as character sets.
```
Comment (spmse) [inline module-5/hydra/README.md]: You could also say "Configurability" or "Parameterization" instead of configurable parameters, which sounds a bit bulky but is indeed fine in that place.

File: module-5/hydra/README.md [hunk 3]
```diff
@@ -1,18 +1,130 @@
...
+- **Logging**: Provides detailed logs of all connection attempts and outcomes.
```
Comment (spmse) [inline module-5/hydra/README.md]: in the best case, add some information if your log output is structured or unstructured --> that way readers can obtain the info if the logs can be processed by other programs or only are there for human debugging and interaction with the logs.

File: module-5/hydra/README.md [hunk 4]
```diff
@@ -1,18 +1,130 @@
...
+
+## Getting Started
+
+To get started with the `hydra` tool, follow these steps:
+
+1. **Clone the Repository**:
+
+```shell
+bash
```
Comment (spmse) [inline module-5/hydra/README.md]: why `bash` here? Or was this a mistake where you couldn't decide between shell and bash for the code-fence block?

File: module-5/hydra/README.md [hunk 5]
```diff
@@ -1,18 +1,130 @@
...
+git clone https://github.com/yourusername/hydra.git
```
Comment (spmse) [inline module-5/hydra/README.md]: if this is your repo and your readme explains how to interact with it, then just go for the real link 👍🏼

File: module-5/hydra/README.md [hunk 6]
```diff
@@ -1,18 +1,130 @@
...
+cd hydra
+```
+
+2. **Install Dependencies**:
+   Ensure you have **paramiko** installed. You can install it using pip:
+
+```shell
+bash
+pip install paramiko
+```
```
Comment (spmse) [inline module-5/hydra/README.md]: Good point, but could be better in the following way:

1. Create a `requirements.txt` that contains all relevant packages that a user needs to have installed to make this work
    1. install from the requirements file, e.g. `pip install -r requirements.txt`
1. add information about setting up an virtual environment --> isolation for dependencies, also provides safety for your other projects because package versions will not confuse each other due to global installations --> what would happen if i have two projects that use paramiko for SSH connections, but the one project uses v2.2.2. and the other uses v1.6. because i need to have legacy support with that lib? 
1. always try to pin versions you use in order to everyone to see what version this worked with for you

File: module-5/hydra/README.md [hunk 7]
```diff
@@ -1,18 +1,130 @@
...
+
+## Usage Examples
+
```
Comment (spmse) [inline module-5/hydra/README.md:47]: I would change the order of elements in that section in the following way:

1. Examples
    1. Brute-Force Attack
    2. Dictionary Attack
    3. All Hydra-Options
    
That way i can clearly see an example before being cognitively overloaded with all options that are offered by your implementation

File: module-5/hydra/README.md [hunk 8]
```diff
@@ -1,18 +1,130 @@
...
+| Option                   | Description                                    | Required |
+| ------------------------ | ---------------------------------------------- | -------- |
+| `-u` <br> `--username`   | Username for SSH login                         | x        |
+| `-s` <br> `--server`     | Server IP address or DNS                       | x        |
+| `-p` <br> `--port`       | Port for SSH connection (default is 22)        |          |
+| `-w` <br> `--wordlist`   | Path to the wordlist for dictionary attack     |          |
+| `-c` <br> `--character`  | Charset for brute force attack                 |          |
+| `--min` <br> `--minimum` | Minimum password length for brute force attack |          |
+| `--max` <br> `--maximum` | Maximum password length for brute force attack |          |
```
Comment (spmse) [inline module-5/hydra/README.md]: in the best case, you would have two distinct columns for the short-version and the long-version of an option. Also, add a column for a default value to illustrate to potential users what parameters have default values, and also obtain the info about the value

| Option | Short-Name | Description | Default value | required |
| :--- | :--- | :--- | :--- | :---: |
| --username | -u | the username to use for the attack | root | x |
| --wordlist| -w | the path to the wordlist with passwords to use for the attack | - | x |


In that example i provide a default for the username, that can be overwritten but i don't need to always provide it with the call to hydra.

File: module-5/hydra/README.md [hunk 9]
```diff
@@ -1,18 +1,130 @@
...
+### Dictionary Attack
+
+To perform a dictionary attack using a wordlist, use the following command:
+
+```shell
+python hydra.py \
+    -u <username> \
+    -s <server> \
+    -p <port> \
+    -w <path_to_wordlist>
```
Comment (spmse) [inline module-5/hydra/README.md]: if this is required for a dictionary attack, you should also mark it required (with a note maybe) to also showcase the requirements and/or edge-cases

File: module-5/hydra/README.md [hunk 10]
```diff
@@ -1,18 +1,130 @@
...
+To perform a brute-force attack without a wordlist, use the following command:
+
+```shell
+python hydra.py \
+    -u <username> \
+    -s <server> \
+    -p <port> \
+    --min <min_length> \
+    --max <max_length> \
+    -c <charset>
```
Comment (spmse) [inline module-5/hydra/README.md]: are these required for your attack type? if so, this should be part of the documentation

File: module-5/hydra/README.md [hunk 11]
```diff
@@ -1,18 +1,130 @@
...
+    -u <username> \
+    -s <server> \
+    -p <port> \
+    -w <path_to_wordlist>
+```
+
+- username: The SSH username.
+- server: The IP address or DNS name of the SSH server.
+- port: The port for SSH connection (default is 22 if not specified).
+- path_to_wordlist: Path to the wordlist file.
```
Comment (spmse) [inline module-5/hydra/README.md]: can be omitted when table is updated, since this section is redundant and the table above can also cover relevant information about e.g. default values or mandatory parameters.

File: module-5/hydra/README.md [hunk 12]
```diff
@@ -1,18 +1,130 @@
...
+    --max <max_length> \
+    -c <charset>
+```
+
+- username: The SSH username.
+- server: The IP address or DNS name of the SSH server.
+- port: The port for SSH connection (default is 22 if not specified).
+- min_length: Minimum length of the password.
+- max_length: Maximum length of the password.
+- charset: Characters to use in the password generation (default is alphanumeric).
```
Comment (spmse) [inline module-5/hydra/README.md]: same as for parameter list above

File: module-5/hydra/README.md [hunk 13]
```diff
@@ -1,18 +1,130 @@
...
+    -s localhost \
+    -p 2222 \
+    --min 1 \
+    --max 4 \
+    -c abc123
+```
+
+## Logging
+
+Logs are written to **hydra.log**. You can check this file to review detailed information about the connection attempts and outcomes.
```
Comment (spmse) [inline module-5/hydra/README.md]: as outlined above, the info if you have structured or unstructered logging would be benefitial for a reader. 

Also, if there is an option to configure the log location I want to know how to.

Where will `hydra.log` be placed? Within the source code folder? Within the directory i am currently in (CWD)? or somewhere else?
What if that location is already occupied by another file named `hydra.log`?

File: module-5/hydra/hydra.py [hunk 14]
```diff
@@ -1 +1,142 @@
...
+
+    Logs each attempt and the outcome. Reports the time taken to complete the attack.
+    """
+    def password_generator():
+        """
+        Generates all possible passwords within the given length and charset.
+        """
+        for length in range(min_len, max_len + 1):
+            for pwd in itertools.product(charset, repeat=length):
+                yield ''.join(pwd)
```
Comment (spmse) [inline module-5/hydra/hydra.py]: in the base case you dont define the function within another function (which is allowed by python in that place, but discouraged due to less readability and it often results in weird hacks.
In the best case define the function in root level of this file and just use it where required.

File: module-5/hydra/hydra.py [hunk 15]
```diff
@@ -1 +1,142 @@
...
+    parser.add_argument('-w', help="Path to the wordlist for dictionary attack")
+    parser.add_argument('--min', type=int, default=1, help="Minimum password length for brute force attack")
+    parser.add_argument('--max', type=int, default=4, help="Maximum password length for brute force attack")
+    parser.add_argument('-c', default=string.ascii_letters + string.digits, help="Charset for brute force attack")
+
+    args = parser.parse_args()
+
+    if args.w:
+        # Dictionary attack
+        log_new_attack_attempt()
```
Comment (spmse) [inline module-5/hydra/hydra.py]: This function should not only log the start of a new attack, but it should also give some information about the attack type we are going to try, dont you think too?

File: module-5/hydra/hydra.py [hunk 16]
```diff
@@ -1 +1,142 @@
...
+        """
+        for length in range(min_len, max_len + 1):
+            for pwd in itertools.product(charset, repeat=length):
+                yield ''.join(pwd)
+    
+    start_time = time.time()
+    log_new_attack_attempt()
+    logger.info("Brute-force attack started" + "\n")
+
+    for password in password_generator():
```
Comment (spmse) [inline module-5/hydra/hydra.py]: code style improvement:
1. initialize your pw list
2. iterate the list and do stuff.

File: module-5/hydra/hydra.py [hunk 17]
```diff
@@ -1 +1,142 @@
...
+        try:
+            with open(args.w, 'r') as wordlist:
+                for password in wordlist:
+                    password = password.strip()
+                    if try_login(args.u, args.s, password, args.p):
+                        logger.info(f"Username: {args.u}, Server: {args.s}, Password found: {password}")
+                        return
+        except FileNotFoundError:
+            logger.error("Wordlist file not found")
+        logger.info("Password not found")
```
Comment (spmse) [inline module-5/hydra/hydra.py]: should also be part of a function, just like the brute-force attack.
That way your main function will remain clean and just control the program flow --> correct

**Review by spmse — COMMENTED** [review]
Looks good to me! 👍🏼 
I've left a few comments for you that should be applied or answered before re-submitting the project for review.
One **general note:** For function definitions that have more than two or three parameters it is commonly a nice tip to add line breaks to your function signature like follows:

```python
def some_func_with_params(
    some: str, 
    longer: str,
    param: str,
    list: List[str], 
    with: str,
    elements: int
) -> bool:
    # implementation starts here or below :)
    pass
```

File: .gitignore
Comment (PascalNehlsen) [reply .gitignore:1]: I only wanted to have changes in the branch that are branch-specific. That's why my top-level .gitignore is not deleted, just moved to a different branch.

File: README.md
Comment (PascalNehlsen) [reply README.md:1]: I only wanted to have changes in the branch that are branch-specific. That's why my top-level README.md is not deleted, just moved to a different branch. I did not know that this was not done. Unfortunately, this is now done in some PRs. I've made a note of it for the future!

File: module-5/hydra/README.md (hunk 1 above)
Comment (PascalNehlsen) [reply module-5/hydra/README.md:1]: I dont know why u are laughin...  :D

File: module-5/hydra/README.md (hunk 2 above)
Comment (PascalNehlsen) [reply module-5/hydra/README.md]: I have changed it to "Configurability". Sounds good to me!

File: module-5/hydra/README.md (hunk 3 above)
Comment (PascalNehlsen) [reply module-5/hydra/README.md]: Its just for the human debugging and interaction. So its unstructured. I changed it.

File: module-5/hydra/README.md (hunk 4 above)
Comment (PascalNehlsen) [reply module-5/hydra/README.md]: Just a typing mistake!

File: module-5/hydra/README.md (hunk 5 above)
Comment (PascalNehlsen) [reply module-5/hydra/README.md]: Changed it to the real link

File: module-5/hydra/README.md (hunk 6 above)
Comment (PascalNehlsen) [reply module-5/hydra/README.md]: - requirements.txt created
- added paramiko with a specific version number
- description of how to install virtual environment added

File: module-5/hydra/README.md (hunk 7 above)
Comment (PascalNehlsen) [reply module-5/hydra/README.md:47]: Order of the examples section changed

File: module-5/hydra/README.md (hunk 8 above)
Comment (PascalNehlsen) [reply module-5/hydra/README.md]: - added two new columns "default value" and "shorthand"
- added default value for username

File: module-5/hydra/README.md (hunk 9 above)
Comment (PascalNehlsen) [reply module-5/hydra/README.md]: added note, that -w <path_to_wordlist> is required for dictionary attack

File: module-5/hydra/README.md (hunk 10 above)
Comment (PascalNehlsen) [reply module-5/hydra/README.md]: Not required, both got default values

File: module-5/hydra/README.md (hunk 11 above)
Comment (PascalNehlsen) [reply module-5/hydra/README.md]: deleted this section, updated the table

File: module-5/hydra/README.md (hunk 12 above)
Comment (PascalNehlsen) [reply module-5/hydra/README.md]: deleted this section, updated the table

File: module-5/hydra/README.md (hunk 13 above)
Comment (PascalNehlsen) [reply module-5/hydra/README.md]: - the unstructured data is now referred to in “features” 
- added information where the hydra.log will be placed
- added case when hydra.log already exists

File: module-5/hydra/hydra.py (hunk 14 above)
Comment (PascalNehlsen) [reply module-5/hydra/hydra.py]: seperate this two functions

File: module-5/hydra/hydra.py (hunk 15 above)
Comment (PascalNehlsen) [reply module-5/hydra/hydra.py]: Isnt the "logger.info("Brute-force attack started" + "\n")" and "logger.info("Dictionary attack started" + "\n")" not the best case?

File: module-5/hydra/hydra.py (hunk 16 above)
Comment (PascalNehlsen) [reply module-5/hydra/hydra.py]: Initialized the password list with "password_list"

File: module-5/hydra/hydra.py (hunk 17 above)
Comment (PascalNehlsen) [reply module-5/hydra/hydra.py]: outsourced the dictionary attack function

File: .gitignore
Comment (spmse) [reply .gitignore:1]: This the same. A branch is just a copy of your repositories current state, e.g. main and add-new-feat have the same data. So if you delete the README for the add-new-feat branch it won't be available for main after merging anymore --> the delete would be propagated

Comment (PascalNehlsen) [reply .gitignore:1]: readded gitignore and top-level readme. Thank you for your explanation!
